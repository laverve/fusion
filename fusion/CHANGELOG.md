# Changelog

## [1.0.1](https://github.com/laverve/fusion/compare/pixi-fusion-v1.0.0...pixi-fusion-v1.0.1) (2024-10-14)


### Bug Fixes

* **fusion:** ignore errors during reinitialization ([1d663d3](https://github.com/laverve/fusion/commit/1d663d39add96dd3ce80f157d97597af310d7ce6))

## [1.0.0](https://github.com/laverve/fusion/compare/pixi-fusion-v1.0.0...pixi-fusion-v1.0.0) (2024-10-13)


### ⚠ BREAKING CHANGES

* creating a new release after cleaning up

### Features

* **pixi.js:** migrates to the lates version + restructuring repo ([b63d3be](https://github.com/laverve/fusion/commit/b63d3bee1a53ee7933b67b8e0574701b051b9186))


### Bug Fixes

* **fusion:** dummy commit to initiate normal release for fusion ([47fc295](https://github.com/laverve/fusion/commit/47fc2954f9a6a6511b92a645bbf7cd53279182ca))


### Miscellaneous Chores

* bump version ([5000aab](https://github.com/laverve/fusion/commit/5000aaba0487d91b51c023333dd07637167cc221))
* release 0.0.10 ([bd91141](https://github.com/laverve/fusion/commit/bd91141158bc1b90cd36315691163b22c681816b))
* release 1.0.0 ([b653a83](https://github.com/laverve/fusion/commit/b653a830d642f1694ab98eca867f5a82c8951eb8))
* release 3.0.0 ([bcb0164](https://github.com/laverve/fusion/commit/bcb0164e2024fab9bca5f217dc54ecea8f6ca4e5))

## [1.0.0](https://github.com/laverve/fusion/compare/fusion-v0.0.10...fusion-v1.0.0) (2024-10-13)


### ⚠ BREAKING CHANGES

* creating a new release after cleaning up

### Features

* **pixi.js:** migrates to the lates version + restructuring repo ([b63d3be](https://github.com/laverve/fusion/commit/b63d3bee1a53ee7933b67b8e0574701b051b9186))


### Bug Fixes

* **fusion:** dummy commit to initiate normal release for fusion ([47fc295](https://github.com/laverve/fusion/commit/47fc2954f9a6a6511b92a645bbf7cd53279182ca))


### Miscellaneous Chores

* bump version ([5000aab](https://github.com/laverve/fusion/commit/5000aaba0487d91b51c023333dd07637167cc221))
* release 0.0.10 ([bd91141](https://github.com/laverve/fusion/commit/bd91141158bc1b90cd36315691163b22c681816b))
* release 1.0.0 ([b653a83](https://github.com/laverve/fusion/commit/b653a830d642f1694ab98eca867f5a82c8951eb8))
* release 3.0.0 ([bcb0164](https://github.com/laverve/fusion/commit/bcb0164e2024fab9bca5f217dc54ecea8f6ca4e5))

## [0.0.10](https://github.com/laverve/fusion/compare/fusion-v3.0.1...fusion-v0.0.10) (2024-10-13)


### Miscellaneous Chores

* release 0.0.10 ([bd91141](https://github.com/laverve/fusion/commit/bd91141158bc1b90cd36315691163b22c681816b))

## [3.0.1](https://github.com/laverve/fusion/compare/fusion-v3.0.0...fusion-v3.0.1) (2024-10-13)


### Bug Fixes

* **fusion:** dummy commit to initiate normal release for fusion ([47fc295](https://github.com/laverve/fusion/commit/47fc2954f9a6a6511b92a645bbf7cd53279182ca))
