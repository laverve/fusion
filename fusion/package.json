{
    "author": "Victor Pishuk <victor.pishuk@gmail.com>",
    "type": "module",
    "bugs": {
        "url": "https://github.com/laverve/fusion/issues"
    },
    "dependencies": {
        "@types/matter-js": "^0.19.7",
        "matter-js": "^0.20.0",
        "pixi.js": "^8.4.1"
    },
    "peerDependencies": {
        "react": "^18.2.0"
    },
    "description": "This module offers a set of common components needed for playing games.",
    "keywords": [
        "game",
        "engine",
        "pixi.js",
        "matter-js",
        "react"
    ],
    "license": "MIT",
    "main": "./dist/src/index.js",
    "types": "./dist/src/index.d.ts",
    "name": "@laverve/fusion",
    "files": [
        "./dist/src/",
        "./src/",
        "package.json",
        "package-lock.json",
        "README.md",
        "LICENSE"
    ],
    "publishConfig": {
        "access": "public"
    },
    "repository": {
        "type": "git",
        "url": "https://github.com/laverve/fusion.git"
    },
    "scripts": {
        "lint": "eslint .",
        "lint:staged": "lint-staged",
        "test": "jest --passWithNoTests",
        "build": "rimraf ./dist && tsc"
    },
    "version": "3.0.0",
    "webpack": "./src/index.ts"
}
