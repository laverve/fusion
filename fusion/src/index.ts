export * from "./game-context";
export * from "./assets-manager";
export * from "./world";
export * from "./camera";
export * from "./layer";
export * from "./assets-manager";
export * from "./game-objects";
export * from "./physics";
export * from "./types";
export * from "./hooks";
