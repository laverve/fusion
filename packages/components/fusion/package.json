{
    "author": "Victor Pishuk <victor.pishuk@gmail.com>",
    "bugs": {
        "url": "https://github.com/laverve/fusion/issues"
    },
    "dependencies": {
        "@types/matter-js": "^0.19.2",
        "matter-js": "^0.19.0",
        "pixi.js": "^7.3.2"
    },
    "peerDependencies": {
        "react": "^18.2.0"
    },
    "description": "This module offers a set of common components needed for playing games.",
    "keywords": [
        "game",
        "engine",
        "pixi.js",
        "matter-js",
        "react"
    ],
    "license": "MIT",
    "main": "./dist/src/index.js",
    "types": "./dist/src/index.d.ts",
    "name": "@laverve/fusion",
    "files": [
        "./dist/src/",
        "./src/",
        "package.json",
        "package-lock.json",
        "README.md",
        "LICENSE"
    ],
    "publishConfig": {
        "access": "public"
    },
    "repository": {
        "type": "git",
        "url": "https://github.com/laverve/fusion.git"
    },
    "scripts": {
        "lint": "eslint --ext .js,.ts,.tsx,.json .",
        "lint:staged": "lint-staged",
        "test": "jest --passWithNoTests",
        "build": "tsc"
    },
    "version": "2.0.2",
    "webpack": "./src/index.ts"
}
