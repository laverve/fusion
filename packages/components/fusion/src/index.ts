export * from "./game-context";
export * from "./physics";
export * from "./stage";
export * from "./game-objects";
export * from "./types";
