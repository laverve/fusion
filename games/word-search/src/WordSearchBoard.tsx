import React, { useState, TouchEvent, MouseEvent, useContext, useEffect, useRef, useMemo } from "react";
import classnames from "classnames";
import { v4 as uuid } from "uuid";
import { parseToRgba, toHex } from "color2k";
import { GameContext, GameStatus } from "@laverve/fusion";
import { WordSearchContext } from "./WordSearch.context";
import { generatePath } from "./lib/path";

export type WordSearchBoardProps = {
    classNames?: Record<"container" | "cell", string>;
    width: number;
    height: number;
};

export const WordSearchBoard: React.FC<WordSearchBoardProps> = ({
    width,
    height,
    classNames = { container: "", cell: "" }
}: WordSearchBoardProps) => {
    const { foundWords, onWordFound, grid, gridCells, words, selectedWordsColors } = useContext(WordSearchContext);
    const { status } = useContext(GameContext);

    const lowerCasedWords = useMemo(() => words.map((w) => w.toLowerCase()), [words]);
    const lowerCasedFoundWords = useMemo(() => foundWords.map((w) => w.toLowerCase()), [foundWords]);

    const containerRef = useRef<HTMLDivElement>(null);
    const [isSelecting, setIsSelecting] = useState<boolean>(false);
    const [selectedCells, setSelectedCells] = useState<Map<string, boolean>>(new Map());
    const [startCellBlock, setStartCellBlock] = useState<HTMLDivElement | null>(null);
    const [endCellBlock, setEndCellBlock] = useState<HTMLDivElement | null>(null);
    const [selectedPathWidth, setSelectedPathWidth] = useState<number>(1);
    const [highlightedPath, setHighlightedPath] = useState("");
    const [selectedPaths, setSelectedPaths] = useState<{ key: string; path: string }[]>([]);

    const selectedWordsColorsWithOpacity = useMemo(
        () =>
            selectedWordsColors.map((color) => {
                const [r, g, b] = parseToRgba(color);
                return toHex(`rgba(${r},${g},${b},0.5)`);
            }),
        [selectedWordsColors]
    );

    useEffect(() => {
        if (status === GameStatus.IN_PROGRESS || status === GameStatus.READY) {
            setSelectedPaths([]);
            setSelectedCells(new Map());
            setStartCellBlock(null);
            setEndCellBlock(null);
            setHighlightedPath("");
        }
    }, [status]);

    useEffect(() => {
        if (containerRef.current) {
            setSelectedPathWidth((containerRef.current.children?.[0]?.getBoundingClientRect()?.width ?? 4) / 1.5);
        }
    }, [containerRef]);

    const updateSelectedPath = ({
        startCell,
        endCell
    }: {
        startCell?: HTMLDivElement | null;
        endCell?: HTMLDivElement | null;
    }) => {
        if (!startCell || !endCell || !containerRef.current) {
            return;
        }

        const startCellBoundingClientRect = startCell.getBoundingClientRect();
        const endCellBoundingClientRect = endCell.getBoundingClientRect();
        const containerClientBoundingRect = containerRef.current.getBoundingClientRect();

        const normalizedStart = {
            x:
                startCellBoundingClientRect.x -
                (containerClientBoundingRect?.left ?? 0) +
                startCellBoundingClientRect.width * 0.5,
            y:
                startCellBoundingClientRect.y -
                (containerClientBoundingRect?.top ?? 0) +
                startCellBoundingClientRect.height * 0.5
        };

        const normalizedEnd = {
            x:
                endCellBoundingClientRect.x -
                (containerClientBoundingRect?.left ?? 0) +
                startCellBoundingClientRect.width * 0.5,
            y:
                endCellBoundingClientRect.y -
                (containerClientBoundingRect?.top ?? 0) +
                startCellBoundingClientRect.height * 0.5
        };

        setHighlightedPath(
            `M ${normalizedStart.x - 1} ${normalizedStart.y - 1} L ${normalizedEnd.x + 1} ${normalizedEnd.y - 1} `
        );
    };

    const updateSelectedCells = (path: string[]) => {
        path.forEach((key) => selectedCells.set(key, true));
        setSelectedCells(selectedCells);
    };

    const isGameInProgress = () => {
        return status === GameStatus.IN_PROGRESS;
    };

    const onMouseUp = () => {
        if (!isGameInProgress()) {
            return;
        }

        const startCell = gridCells.get(startCellBlock?.getAttribute("data-key") ?? "");
        const endCell = gridCells.get(endCellBlock?.getAttribute("data-key") ?? "");
        const path = generatePath(startCell, endCell);

        if (path.length > 0) {
            const word = path
                .map((key) => gridCells.get(key)?.label.toLowerCase())
                .filter((c) => c)
                .join("");

            const foundIdx = lowerCasedWords.findIndex((w) => w === word);
            const foundEarlierIdx = lowerCasedFoundWords.findIndex((w) => w === word);

            if (foundIdx > -1 && foundEarlierIdx < 0) {
                onWordFound(words[foundIdx]);
                updateSelectedCells(path);
                setSelectedPaths([...selectedPaths, { key: uuid(), path: highlightedPath }]);
            }
        }

        setIsSelecting(false);
        setEndCellBlock(null);
    };

    const onInteractionStart = (event: TouchEvent<HTMLDivElement> | MouseEvent<HTMLDivElement>) => {
        if (!isGameInProgress()) {
            return;
        }

        const target = event.target as HTMLDivElement;
        const key = target.getAttribute("data-key");
        if (key && containerRef.current) {
            setIsSelecting(true);
            setStartCellBlock(target);
            updateSelectedPath({
                startCell: target,
                endCell: target
            });
        }
    };

    const updateSelectedPathFromEventTarget = (target: HTMLDivElement) => {
        if (!isSelecting || !target.getAttribute("data-key")) {
            return;
        }

        const key = target.getAttribute("data-key");

        if (!key) {
            return;
        }

        const cell = gridCells.get(key);

        if (!cell) {
            return;
        }

        updateSelectedPath({
            startCell: startCellBlock,
            endCell: target
        });

        setEndCellBlock(target);
    };

    const onMouseMove = (event: MouseEvent<HTMLDivElement>) => {
        if (!isGameInProgress()) {
            return;
        }

        updateSelectedPathFromEventTarget(event.target as HTMLDivElement);
    };

    const onTouchMove = (event: TouchEvent<HTMLDivElement>) => {
        if (!isGameInProgress()) {
            return;
        }

        const cells = event.currentTarget.children;
        const { clientX, clientY } = event.touches[0];

        for (let i = 0; i < cells.length; i++) {
            const cell = cells.item(i);
            if (cell) {
                const bbox = cell.getBoundingClientRect();
                if (
                    clientX >= bbox.left &&
                    clientX <= bbox.left + bbox.width &&
                    clientY >= bbox.top &&
                    clientY <= bbox.top + bbox.height
                ) {
                    updateSelectedPathFromEventTarget(cell as HTMLDivElement);
                    break;
                }
            }
        }
    };

    useEffect(() => {
        document.addEventListener("mouseup", onMouseUp);
        document.addEventListener("touchend", onMouseUp);
        return () => {
            document.removeEventListener("mouseup", onMouseUp);
            document.removeEventListener("touchend", onMouseUp);
        };
    }, [onMouseUp]);

    const highlightColor = selectedWordsColorsWithOpacity[selectedPaths.length % selectedWordsColorsWithOpacity.length];

    return (
        <svg
            xmlns="http://www.w3.org/2000/svg"
            viewBox={`0 0 ${width} ${height}`}
            className="touch-none"
            style={{
                width: "100%"
            }}
        >
            <foreignObject x="0" y="0" width={width} height={height}>
                <div
                    data-testid="testid-wordsearch-board"
                    className={classnames(
                        "bg-divider",
                        "border-divider",
                        "p-1",
                        "gap-1",
                        classNames?.container ?? "",
                        "grid",
                        "box-border",
                        `grid-cols-${grid?.[0]?.length || 1}`,
                        `grid-rows-${grid?.length || 1}`,
                        { "blur-sm": !isGameInProgress() }
                    )}
                    style={{
                        height,
                        width,
                        borderWidth: "initial !important"
                    }}
                    onMouseDown={onInteractionStart}
                    onMouseMove={onMouseMove}
                    onTouchStart={onInteractionStart}
                    onTouchMove={onTouchMove}
                    role="button"
                    tabIndex={0}
                    ref={containerRef}
                >
                    {grid.flatMap((row, rowId) =>
                        row.map((char, colId) => (
                            <div
                                key={gridCells.get(`${rowId}-${colId}`)?.key}
                                data-testid={`testid-${rowId}-${colId}`}
                                className={classnames(
                                    "uppercase",
                                    "cursor-pointer",
                                    "bg-foreground",
                                    classNames.cell,
                                    "flex",
                                    "items-center",
                                    "justify-center"
                                )}
                                data-key={`${rowId}-${colId}`}
                            >
                                {char}
                            </div>
                        ))
                    )}
                </div>
            </foreignObject>
            {selectedPaths.map(({ key, path }, i) => (
                <path
                    key={key}
                    d={path}
                    stroke={selectedWordsColorsWithOpacity[i % selectedWordsColorsWithOpacity.length]}
                    strokeWidth={selectedPathWidth}
                    strokeLinejoin="round"
                    strokeLinecap="round"
                    pointerEvents="none"
                />
            ))}
            {isSelecting && (
                <path
                    d={highlightedPath}
                    stroke={highlightColor}
                    strokeWidth={selectedPathWidth}
                    strokeLinejoin="round"
                    strokeLinecap="round"
                    pointerEvents="none"
                />
            )}
        </svg>
    );
};
