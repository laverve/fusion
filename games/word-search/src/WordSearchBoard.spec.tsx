/**
 * @jest-environment jsdom
 */

import { describe, it, expect } from "@jest/globals";
import React, { useContext } from "react";

import snapshotDiff from "snapshot-diff";

import { render, fireEvent, waitFor } from "@testing-library/react";
import { GameContextProvider, GameContext } from "@laverve/fusion";

import { WordSearchContextProvider } from "./WordSearch.context";
import { WordSearchBoard, WordSearchBoardProps } from "./WordSearchBoard";

let stopSpy: jest.SpyInstance;

export const TestGameplay: React.FC = () => {
    const context = useContext(GameContext);
    stopSpy = jest.spyOn(context, "stop");
    stopSpy.mockImplementationOnce(() => {});

    return (
        <>
            <button
                type="button"
                data-testid="test-start"
                onClick={() => {
                    context.start();
                }}
            >
                Start
            </button>
            <button
                type="button"
                data-testid="test-stop"
                onClick={() => {
                    context.stop();
                }}
            >
                Stop
            </button>
        </>
    );
};

describe("WordSearchBoard", () => {
    const WORDS = ["dig", "boat"];
    const GRID = [
        ["b", "d", "i", "g"],
        ["d", "o", "g", "o"],
        ["u", "c", "a", "t"],
        ["x", "y", "z", "t"]
    ];

    const renderComponentUderTest = ({
        props = { width: 300, height: 300 },
        timeout = 0,
        selectedWordsColors = ["black"]
    }: {
        props?: WordSearchBoardProps;
        timeout?: number;
        selectedWordsColors?: string[];
    } = {}) => {
        return render(
            <GameContextProvider timeout={timeout}>
                <WordSearchContextProvider grid={GRID} words={WORDS} selectedWordsColors={selectedWordsColors}>
                    <TestGameplay />
                    <WordSearchBoard {...props} />
                </WordSearchContextProvider>
            </GameContextProvider>
        );
    };

    it("should render when game is not started", () => {
        const component = renderComponentUderTest();
        expect(component.asFragment()).toMatchSnapshot();
    });

    it("should disregard selected words before game is started", () => {
        const component = renderComponentUderTest();
        const initialSnapshot = component.asFragment();

        fireEvent.mouseDown(component.getByTestId("testid-0-0"), {
            bubbles: true
        });

        fireEvent.mouseMove(component.getByTestId("testid-1-1"), {
            bubbles: true
        });

        fireEvent.mouseMove(component.getByTestId("testid-2-2"), {
            bubbles: true
        });
        expect(snapshotDiff(initialSnapshot, component.asFragment())).toMatchSnapshot();
    });

    it("should unblur board once game is started", () => {
        const component = renderComponentUderTest();
        const initialSnapshot = component.asFragment();

        fireEvent.click(component.getByTestId("test-start"));
        expect(snapshotDiff(initialSnapshot, component.asFragment())).toMatchSnapshot();
    });

    it("should highlight path that is being selected", () => {
        const component = renderComponentUderTest();
        const initialSnapshot = component.asFragment();

        fireEvent.click(component.getByTestId("test-start"));

        fireEvent.mouseDown(component.getByTestId("testid-0-0"), {
            bubbles: true
        });

        fireEvent.mouseMove(component.getByTestId("testid-1-1"), {
            bubbles: true
        });

        fireEvent.mouseMove(component.getByTestId("testid-2-2"), {
            bubbles: true
        });
        expect(snapshotDiff(initialSnapshot, component.asFragment())).toMatchSnapshot();
    });

    it("should highlight selected word and path that is being selected", () => {
        const component = renderComponentUderTest();
        const initialSnapshot = component.asFragment();

        fireEvent.click(component.getByTestId("test-start"));

        fireEvent.mouseDown(component.getByTestId("testid-0-0"), {
            bubbles: true
        });

        fireEvent.mouseMove(component.getByTestId("testid-1-1"), {
            bubbles: true
        });

        fireEvent.mouseMove(component.getByTestId("testid-2-2"), {
            bubbles: true
        });

        fireEvent.mouseMove(component.getByTestId("testid-3-3"), {
            bubbles: true
        });

        fireEvent.mouseUp(component.getByTestId("testid-3-3"), {
            bubbles: true
        });

        fireEvent.mouseDown(component.getByTestId("testid-0-1"), {
            bubbles: true
        });

        fireEvent.mouseMove(component.getByTestId("testid-0-2"), {
            bubbles: true
        });
        expect(snapshotDiff(initialSnapshot, component.asFragment())).toMatchSnapshot();
    });

    it("should stop the game once all words are selected", async () => {
        const component = renderComponentUderTest();

        fireEvent.click(component.getByTestId("test-start"));

        expect(stopSpy).not.toHaveBeenCalled();

        [
            ["0-0", "1-1", "2-2", "3-3"],
            ["0-1", "0-2", "0-3"],
            ["1-0", "2-0", "3-0"]
        ].forEach((arr) => {
            fireEvent.mouseDown(component.getByTestId(`testid-${arr[0]}`), {
                bubbles: true
            });

            for (let i = 1; i < arr.length; i++) {
                fireEvent.mouseMove(component.getByTestId(`testid-${arr[i]}`), {
                    bubbles: true
                });
            }

            fireEvent.mouseUp(component.getByTestId(`testid-${arr[arr.length - 1]}`), {
                bubbles: true
            });
        });

        await waitFor(() => expect(stopSpy).toHaveBeenCalled(), { timeout: 10000 });
    });
});
