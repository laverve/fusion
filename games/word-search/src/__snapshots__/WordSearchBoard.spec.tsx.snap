// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WordSearchBoard should disregard selected words before game is started 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`WordSearchBoard should highlight path that is being selected 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -25,11 +25,11 @@
          width="300"
          x="0"
          y="0"
        >
          <div
-           class="bg-divider border-divider p-1 gap-1 grid box-border grid-cols-4 grid-rows-4 blur-sm"
+           class="bg-divider border-divider p-1 gap-1 grid box-border grid-cols-4 grid-rows-4"
            data-testid="testid-wordsearch-board"
            role="button"
            style="height: 300px; width: 300px;"
            tabindex="0"
          >
@@ -145,8 +145,16 @@
            >
              t
            </div>
          </div>
        </foreignobject>
+       <path
+         d="M -1 -1 L 1 -1 "
+         pointer-events="none"
+         stroke="#00000080"
+         stroke-linecap="round"
+         stroke-linejoin="round"
+         stroke-width="0"
+       />
      </svg>
    </div>
  </DocumentFragment>"
`;

exports[`WordSearchBoard should highlight selected word and path that is being selected 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -25,11 +25,11 @@
          width="300"
          x="0"
          y="0"
        >
          <div
-           class="bg-divider border-divider p-1 gap-1 grid box-border grid-cols-4 grid-rows-4 blur-sm"
+           class="bg-divider border-divider p-1 gap-1 grid box-border grid-cols-4 grid-rows-4"
            data-testid="testid-wordsearch-board"
            role="button"
            style="height: 300px; width: 300px;"
            tabindex="0"
          >
@@ -145,8 +145,24 @@
            >
              t
            </div>
          </div>
        </foreignobject>
+       <path
+         d="M -1 -1 L 1 -1 "
+         pointer-events="none"
+         stroke="#00000080"
+         stroke-linecap="round"
+         stroke-linejoin="round"
+         stroke-width="0"
+       />
+       <path
+         d="M -1 -1 L 1 -1 "
+         pointer-events="none"
+         stroke="#00000080"
+         stroke-linecap="round"
+         stroke-linejoin="round"
+         stroke-width="0"
+       />
      </svg>
    </div>
  </DocumentFragment>"
`;

exports[`WordSearchBoard should render when game is not started 1`] = `
<DocumentFragment>
  <div
    data-overlay-container="true"
  >
    <button
      data-testid="test-start"
      type="button"
    >
      Start
    </button>
    <button
      data-testid="test-stop"
      type="button"
    >
      Stop
    </button>
    <svg
      class="touch-none"
      style="width: 100%;"
      viewBox="0 0 300 300"
      xmlns="http://www.w3.org/2000/svg"
    >
      <foreignobject
        height="300"
        width="300"
        x="0"
        y="0"
      >
        <div
          class="bg-divider border-divider p-1 gap-1 grid box-border grid-cols-4 grid-rows-4 blur-sm"
          data-testid="testid-wordsearch-board"
          role="button"
          style="height: 300px; width: 300px;"
          tabindex="0"
        >
          <div
            class="uppercase cursor-pointer bg-foreground flex items-center justify-center"
            data-key="0-0"
            data-testid="testid-0-0"
          >
            b
          </div>
          <div
            class="uppercase cursor-pointer bg-foreground flex items-center justify-center"
            data-key="0-1"
            data-testid="testid-0-1"
          >
            d
          </div>
          <div
            class="uppercase cursor-pointer bg-foreground flex items-center justify-center"
            data-key="0-2"
            data-testid="testid-0-2"
          >
            i
          </div>
          <div
            class="uppercase cursor-pointer bg-foreground flex items-center justify-center"
            data-key="0-3"
            data-testid="testid-0-3"
          >
            g
          </div>
          <div
            class="uppercase cursor-pointer bg-foreground flex items-center justify-center"
            data-key="1-0"
            data-testid="testid-1-0"
          >
            d
          </div>
          <div
            class="uppercase cursor-pointer bg-foreground flex items-center justify-center"
            data-key="1-1"
            data-testid="testid-1-1"
          >
            o
          </div>
          <div
            class="uppercase cursor-pointer bg-foreground flex items-center justify-center"
            data-key="1-2"
            data-testid="testid-1-2"
          >
            g
          </div>
          <div
            class="uppercase cursor-pointer bg-foreground flex items-center justify-center"
            data-key="1-3"
            data-testid="testid-1-3"
          >
            o
          </div>
          <div
            class="uppercase cursor-pointer bg-foreground flex items-center justify-center"
            data-key="2-0"
            data-testid="testid-2-0"
          >
            u
          </div>
          <div
            class="uppercase cursor-pointer bg-foreground flex items-center justify-center"
            data-key="2-1"
            data-testid="testid-2-1"
          >
            c
          </div>
          <div
            class="uppercase cursor-pointer bg-foreground flex items-center justify-center"
            data-key="2-2"
            data-testid="testid-2-2"
          >
            a
          </div>
          <div
            class="uppercase cursor-pointer bg-foreground flex items-center justify-center"
            data-key="2-3"
            data-testid="testid-2-3"
          >
            t
          </div>
          <div
            class="uppercase cursor-pointer bg-foreground flex items-center justify-center"
            data-key="3-0"
            data-testid="testid-3-0"
          >
            x
          </div>
          <div
            class="uppercase cursor-pointer bg-foreground flex items-center justify-center"
            data-key="3-1"
            data-testid="testid-3-1"
          >
            y
          </div>
          <div
            class="uppercase cursor-pointer bg-foreground flex items-center justify-center"
            data-key="3-2"
            data-testid="testid-3-2"
          >
            z
          </div>
          <div
            class="uppercase cursor-pointer bg-foreground flex items-center justify-center"
            data-key="3-3"
            data-testid="testid-3-3"
          >
            t
          </div>
        </div>
      </foreignobject>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`WordSearchBoard should unblur board once game is started 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -25,11 +25,11 @@
          width="300"
          x="0"
          y="0"
        >
          <div
-           class="bg-divider border-divider p-1 gap-1 grid box-border grid-cols-4 grid-rows-4 blur-sm"
+           class="bg-divider border-divider p-1 gap-1 grid box-border grid-cols-4 grid-rows-4"
            data-testid="testid-wordsearch-board"
            role="button"
            style="height: 300px; width: 300px;"
            tabindex="0"
          >"
`;
